src/Confluent.Kafka/Impl/SafeKafkaHandle.cs

```diff
@@ -224,91 +225,51 @@ private IntPtr marshalHeaders(IEnumerable<Header> headers)
             return headersPtr;
         }
 
-        internal ErrorCode Produce(
-            string topic, 
-            byte[] val, int valOffset, int valLength,
-            byte[] key, int keyOffset, int keyLength,
+        internal unsafe ErrorCode Produce(
+            string topic,
+            ReadOnlySpan<byte> val,
+            ReadOnlySpan<byte> key,
             int partition,
             long timestamp,
             IEnumerable<Header> headers,
             IntPtr opaque,
             bool blockIfQueueFull)
         {
-            var pValue = IntPtr.Zero;
-            var pKey = IntPtr.Zero;
-
-            var gchValue = default(GCHandle);
-            var gchKey = default(GCHandle);
-
-            if (val == null)
-            {
-                if (valOffset != 0 || valLength != 0)
-                {
-                    throw new ArgumentException("valOffset and valLength parameters must be 0 when producing null values.");
-                }
-            }
-            else
+            fixed (byte* valPtr = &MemoryMarshal.GetReference(val))
+            fixed (byte* keyPtr = &MemoryMarshal.GetReference(key))
             {
-                gchValue = GCHandle.Alloc(val, GCHandleType.Pinned);
-                pValue = Marshal.UnsafeAddrOfPinnedArrayElement(val, valOffset);
-            }
+                IntPtr headersPtr = marshalHeaders(headers);
 
-            if (key == null)
-            {
-                if (keyOffset != 0 || keyLength != 0)
+                try
                 {
-                    throw new ArgumentException("keyOffset and keyLength parameters must be 0 when producing null key values.");
-                }
-            }
-            else
-            {
-                gchKey = GCHandle.Alloc(key, GCHandleType.Pinned);
-                pKey = Marshal.UnsafeAddrOfPinnedArrayElement(key, keyOffset);
-            }
-
-            IntPtr headersPtr = marshalHeaders(headers);
+                    var errorCode = Librdkafka.producev(
+                        handle,
+                        topic,
+                        partition,
+                        (IntPtr)(MsgFlags.MSG_F_COPY | (blockIfQueueFull ? MsgFlags.MSG_F_BLOCK : 0)),
+                        Unsafe.AsRef<byte>(valPtr), (UIntPtr)val.Length,
```

Comment:
I changed the signature of this function from `IntPtr` to `in byte`. If that doesn't make sense, I can change it back, and
`Unsafe.AsRef<byte>(valPtr)`
can be replaced with
`(IntPtr)valPtr`